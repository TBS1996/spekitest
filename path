[front]
text = "hey whats up?"

[back]
text = """

"""

[meta]
id = "550e8400-e29b-41d4-a716-446655440000"
dependencies = []
dependents = []
suspended = false
finished = true
tags = []

[[history]]
timestamp = 0
grade = "none"
time_spent = 0

[[history]]
timestamp = 0
grade = "none"
time_spent = 0

[[history]]
timestamp = 0
grade = "none"
time_spent = 0

[[history]]
timestamp = 1683982056
grade = "none"
time_spent = 0

[[history]]
timestamp = 1683982064
grade = "none"
time_spent = 0

[[history]]
timestamp = 1683982074
grade = "none"
time_spent = 0

[[history]]
timestamp = 1683982157
grade = "none"
time_spent = 0

[[history]]
timestamp = 1683982707
grade = "none"
time_spent = 0
