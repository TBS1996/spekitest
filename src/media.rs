use std::{fs::File, io::BufReader, path::PathBuf};

use rodio::Source;
use serde::{Deserialize, Deserializer, Serialize};

use crate::get_media_path;

#[derive(Deserialize, Clone, Serialize, Debug, Default)]
pub struct AudioSource {
    #[serde(default)]
    #[serde(rename = "audio_local")]
    local_name: Option<String>,
    #[serde(default)]
    #[serde(rename = "audio_url")]
    url_backup: Option<String>,
}

impl AudioSource {
    pub fn play_audio(&mut self) -> Option<()> {
        let path = self.get_path().unwrap();
        let (_stream, stream_handle) = rodio::OutputStream::try_default().unwrap();

        // Load a sound from a file
        let file = BufReader::new(File::open(path).unwrap());
        let source = rodio::Decoder::new(file).unwrap();

        // Play the sound
        stream_handle.play_raw(source.convert_samples()).unwrap();

        // The sound plays in a separate audio thread, so we need to keep the main thread alive while it's playing.
        std::thread::sleep(std::time::Duration::from_secs(10));

        Some(())
    }
}

impl GetMedia for AudioSource {
    fn local_name(&self) -> Option<String> {
        self.local_name.clone()
    }

    fn url_backup(&self) -> Option<String> {
        self.url_backup.clone()
    }

    fn update_local(&mut self, name: &str) {
        self.local_name = Some(name.into());
    }
}

pub trait GetMedia {
    fn local_name(&self) -> Option<String>;
    fn url_backup(&self) -> Option<String>;
    fn update_local(&mut self, name: &str);

    fn local_name_as_path(&self) -> Option<PathBuf> {
        let local_file = self.local_name()?;
        Some(get_media_path().join(local_file))
    }

    fn get_local_path(&self) -> Option<PathBuf> {
        let local_name = self.local_name_as_path()?;
        if local_name.exists() {
            return Some(local_name);
        }
        None
    }

    fn download_media(&mut self) -> Option<PathBuf> {
        let url = self.url_backup()?;

        let response = ureq::get(&url).call().unwrap();

        let fname = url.rsplit('/').next().unwrap();
        let path = get_media_path().join(fname);

        let mut out = std::fs::File::create(&path).unwrap();

        let mut reader = response.into_reader();
        std::io::copy(&mut reader, &mut out).ok()?;

        self.update_local(fname);

        Some(path)
    }

    fn get_path(&mut self) -> Option<PathBuf> {
        self.get_local_path().or_else(|| self.download_media())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_download_media() {
        let mut media = AudioSource {
            local_name: None,
            url_backup: Some("https://tatoeba.org/en/audio/download/269099".into()),
        };
        let path = media.get_path().unwrap();
        assert_eq!(media.local_name.unwrap(), "269099");
        std::fs::remove_file(path).unwrap();
    }
}
