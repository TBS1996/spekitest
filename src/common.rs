
use crate::{Id, PATH};
use std::fs;
use std::io;
use std::path::Path;
use std::path::PathBuf;
use std::time::{Duration, SystemTime};

pub fn current_time() -> Duration {
    let now = SystemTime::now()
        .duration_since(SystemTime::UNIX_EPOCH)
        .expect("Time went backwards");
    now
}

// Represent the category that a card is in, can be nested
#[derive(Debug, Clone, Default, PartialEq)]
pub struct Category(pub Vec<String>);

impl Category {
    pub fn load_all() -> io::Result<Vec<Category>> {
        let root = Path::new(PATH);
        let mut folders = Vec::new();
        Self::collect_folders_inner(root, root, &mut folders)?;
        folders.push(Category::default());
        Ok(folders)
    }

    pub fn append(mut self, category: &str) -> Self {
        self.0.push(category.into());
        self
    }

    fn collect_folders_inner(
        root: &Path,
        current: &Path,
        folders: &mut Vec<Category>,
    ) -> io::Result<()> {
        if current.is_dir() {
            for entry in fs::read_dir(current)? {
                let entry = entry?;
                let path = entry.path();
                if path.is_dir() {
                    // Compute the relative path from root to the current directory
                    let rel_path = path
                        .strip_prefix(root)
                        .expect("Failed to compute relative path")
                        .components()
                        .map(|c| c.as_os_str().to_string_lossy().into_owned())
                        .collect::<Vec<String>>();
                    folders.push(Self(rel_path));
                    Self::collect_folders_inner(root, &path, folders)?;
                }
            }
        }
        Ok(())
    }

    pub fn joined(&self) -> String {
        self.0.join("/")
    }

    pub fn from_card_path(path: &Path) -> Self {
        let root = Path::new(PATH);
        let path = path.strip_prefix(root).unwrap();

        let mut x = path
            .components()
            .map(|c| c.as_os_str().to_string_lossy().into_owned())
            .collect::<Vec<String>>();
        x.pop();
        Category(x)
    }

    pub fn from_string(s: String) -> Self {
        let vec = s.split("/").map(|s| s.to_string()).collect();
        Self(vec)
    }

    pub fn as_path(&self) -> PathBuf {
        let categories = self.0.join("/");
        let path = format!("{}{}", PATH, categories);
        PathBuf::from(path)
    }
    pub fn as_path_with_id(&self, id: Id) -> PathBuf {
        let mut folder = self.as_path();
        folder = folder.join(id.to_string());
        folder.set_extension("toml");
        folder
    }
}

pub mod serde_duration_as_secs {
    use serde::{Deserialize, Deserializer, Serializer};
    use std::time::Duration;

    pub fn serialize<S>(duration: &Duration, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let secs = duration.as_secs();
        serializer.serialize_u64(secs)
    }

    pub fn deserialize<'de, D>(deserializer: D) -> Result<Duration, D::Error>
    where
        D: Deserializer<'de>,
    {
        let secs = u64::deserialize(deserializer)?;
        Ok(Duration::from_secs(secs))
    }
}

#[cfg(test)]
mod tests {
    use uuid::uuid;

    use super::*;

    #[test]
    fn test_load_all() {
        let root = Path::new(PATH);
        let mut folders = vec![];
        Category::collect_folders_inner(&root, &root, &mut folders).unwrap();

        insta::assert_debug_snapshot!(folders);
    }

    #[test]
    fn test_joined() {
        let category = Category(vec!["foo".into(), "bar".into()]);
        let joined = category.joined();
        insta::assert_debug_snapshot!(joined);
    }

    #[test]
    fn test_from_card_path() {
        let card_path = "./testing/maths/calculus/491f8b92-c943-4c4b-b7bf-f7d483208eb0.toml";
        let card_path = Path::new(card_path);
        let x = Category::from_card_path(card_path);
        insta::assert_debug_snapshot!(x);
    }

    #[test]
    fn test_as_path() {
        let category = Category(vec!["foo".into(), "bar".into()]);
        let x = category.as_path();
        insta::assert_debug_snapshot!(x);
    }

    #[test]
    fn test_as_path_with_id() {
        let id = uuid!("8bc35fe2-f02b-4633-8f1b-306eb4e09cd2");
        let category = Category(vec!["foo".into(), "bar".into()]);
        let x = category.as_path_with_id(id);
        insta::assert_debug_snapshot!(x);
    }
}
