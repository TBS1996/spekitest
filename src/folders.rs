use uuid::Uuid;

use crate::common::Category;
use crate::frontend;
use crate::{Conn, Id};

pub fn get_cards_from_category(category: &Category) -> Vec<Id> {
    let directory = category.as_path();
    let mut toml_files = Vec::new();

    for entry in std::fs::read_dir(directory).unwrap() {
        let entry = entry.unwrap();
        let path = entry.path();

        if path.is_file() && path.extension().and_then(|s| s.to_str()) == Some("toml") {
            if let Some(filename) = path.file_stem().and_then(|s| s.to_str()) {
                let id = Uuid::parse_str(filename).unwrap();
                toml_files.push(id);
            }
        }
    }
    toml_files
}

pub fn get_all_cards() -> Vec<Id> {
    let cats = Category::load_all().unwrap();
    let mut cards = vec![];

    for cat in &cats {
        cards.extend(get_cards_from_category(cat));
    }
    cards
}

pub fn review_card_in_directory(conn: &Conn, category: &Category) {
    let cards = get_all_cards();
    frontend::review_cards(conn, cards, category);
}

pub fn search_for_id(id: Id) -> Option<Category> {
    let folders = Category::load_all().unwrap();

    for folder in folders {
        let full_path = folder.as_path_with_id(id);

        if full_path.exists() {
            return Some(folder);
        }
    }
    None
}

pub fn create_category(category: &Category) -> Category {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    let input = normalize_category_name(&input);
    let category = category.clone().append(&input);
    let path = category.as_path();
    std::fs::create_dir(path).unwrap();
    category
}

pub fn normalize_category_name(input: &str) -> String {
    let mut normalized = String::with_capacity(input.len());

    for c in input.chars() {
        if c.is_ascii_alphanumeric() {
            normalized.push(c);
        } else if c == ' ' {
            normalized.push(c);
        }
    }

    normalized
}
