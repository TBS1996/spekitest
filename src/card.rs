use serde::{Deserialize, Serialize};
use std::ffi::OsStr;
use std::fs;
use std::fs::read_to_string;
use std::io::{self};
use std::path::{Path, PathBuf};
use std::process::Command;
use std::time::Duration;
use uuid::Uuid;

use crate::common::Category;
use crate::folders::{get_all_cards_full, get_category_from_id_from_fs};
use crate::media::AudioSource;
use crate::{common::current_time, Id};

pub struct CardFileData {
    pub file_name: String,
    pub category: Category,
    pub last_modified: Duration,
}

pub struct CardWithFileData(pub Card, pub CardFileData);

impl CardWithFileData {
    pub fn get_full_card_from_id(id: Id) -> Option<Self> {
        let cards = get_all_cards_full();

        cards.into_iter().find(|card| card.0.meta.id == id)
    }

    pub fn into_card(self) -> Card {
        self.0
    }

    pub fn into_cards(v: Vec<Self>) -> Vec<Card> {
        v.into_iter().map(|c| c.into_card()).collect()
    }

    pub fn full_path(&self) -> PathBuf {
        let mut path = self.1.category.as_path().join(self.1.file_name.clone());
        path.set_extension("toml");
        path
    }
}

#[derive(Deserialize, Serialize, Debug, Default, Clone)]
pub struct Card {
    pub front: Side,
    pub back: Side,
    pub meta: Meta,
    #[serde(default)]
    pub history: Vec<Review>,
}

// public
impl Card {
    pub fn new(front: Side, back: Side, meta: Meta) -> Self {
        Card {
            front,
            back,
            meta,
            history: Vec::new(),
        }
    }

    pub fn new_simple(front: String, back: String) -> Self {
        Card {
            front: Side {
                text: front,
                ..Default::default()
            },
            back: Side {
                text: back,
                ..Default::default()
            },
            ..Default::default()
        }
    }

    pub fn recall_rate(&self) -> Option<f32> {
        let days_passed = self.days_since_last_review()?;
        let stability = self.meta.stability?;
        Some(Self::calculate_strength(days_passed, stability))
    }

    fn calculate_strength(days_passed: f32, stability: f32) -> f32 {
        let base: f32 = 0.9;
        (base.ln() * days_passed / stability).exp()
    }

    pub fn days_since_last_review(&self) -> Option<f32> {
        let last_unix = self.history.last()?.timestamp;
        let current_unix = current_time();
        Some((current_unix - last_unix).as_secs_f32() / 86400.)
    }

    pub fn is_ready_for_review(&self) -> bool {
        match (self.meta.stability, self.days_since_last_review()) {
            (Some(stability), Some(last_review_time)) => {
                self.meta.finished
                    && !self.meta.suspended
                    && last_review_time > (1. / 1440.) // Lets not review if its less than a minute since last time
                    && stability < last_review_time
            }
            (_, _) => false,
        }
    }

    pub fn load_from_id(id: Id) -> Option<Self> {
        let card = CardWithFileData::get_full_card_from_id(id)?;
        Some(card.into_card())
    }

    pub fn save_card(self, incoming_category: Option<Category>) {
        let incoming_category = incoming_category
            .or_else(|| get_category_from_id_from_fs(self.meta.id))
            .unwrap_or(Category(vec![]));

        let id = self.meta.id;

        self.save_card_to_toml(&incoming_category).unwrap();
    }

    pub fn save_card_to_toml(&self, category: &Category) -> Result<PathBuf, toml::ser::Error> {
        let toml = toml::to_string(&self).unwrap();
        std::fs::create_dir_all(category.as_path()).unwrap();
        let path = category
            .as_path()
            .join(self.front.text.clone())
            .with_extension("toml");

        let _ = std::fs::write(&path, toml);
        Ok(path)
    }

    // The closure takes a Card and returns a Result.
    // This allows it to handle errors that might occur during processing.
    pub fn _process_cards<F>(dir: &Path, func: &mut F) -> io::Result<()>
    where
        F: FnMut(Card, &Category) -> io::Result<()>,
    {
        if dir.is_dir() {
            let entries = fs::read_dir(dir)?;
            for entry in entries {
                let entry = entry?;
                let path = entry.path();
                if path.is_dir() {
                    Self::_process_cards(&path, func)?;
                } else if path.extension() == Some(OsStr::new("toml")) {
                    let card = Self::parse_toml_to_card(&path).unwrap(); // Assuming parse_toml_to_card returns Result<Card, io::Error>
                    let category = Category::_from_card_path(&path);
                    func(card, &category)?;
                }
            }
        }
        Ok(())
    }

    pub fn _create_new(front: &str, back: &str, category: &Category) -> Id {
        let card = Card {
            front: Side {
                text: front.into(),
                ..Default::default()
            },
            back: Side {
                text: back.into(),
                ..Default::default()
            },
            ..Default::default()
        };

        let id = card.meta.id;

        card.save_card_to_toml(category).unwrap();
        id
    }

    fn new_stability(grade: Grade, time_passed: Option<f32>) -> f32 {
        grade.get_factor() * time_passed.unwrap_or(1.)
    }

    pub fn new_review(&mut self, grade: Grade, category: &Category) {
        let review = Review::new(grade.clone());
        self.history.push(review);
        self.meta.stability = Some(Self::new_stability(grade, self.meta.stability));
        self.save_card_to_toml(category).unwrap();
    }

    pub fn parse_toml_to_card(file_path: &Path) -> Result<Card, toml::de::Error> {
        let content = read_to_string(file_path).expect("Could not read the TOML file");
        let mut card: Card = toml::from_str(&content)?;
        card.history.sort_by(|a, b| a.timestamp.cmp(&b.timestamp));
        Ok(card)
    }
}

#[derive(Deserialize, Serialize, Debug, Default, Clone)]
#[serde(rename_all = "lowercase")]
pub enum Grade {
    // no recall, not even when you saw the answer.
    #[default]
    None,
    // no recall, but you remember the answer when you read it.
    Late,
    // struggled but you got the answer right or somewhat right/
    Some,
    // no hesitation, perfect recall.
    Perfect,
}

impl Grade {
    pub fn get_factor(&self) -> f32 {
        match self {
            Grade::None => 0.1,
            Grade::Late => 0.25,
            Grade::Some => 2.,
            Grade::Perfect => 3.,
        }
    }
}

impl std::str::FromStr for Grade {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "1" => Ok(Self::None),
            "2" => Ok(Self::Late),
            "3" => Ok(Self::Some),
            "4" => Ok(Self::Perfect),
            _ => Err(()),
        }
    }
}

use crate::common::serde_duration_as_secs;

#[derive(Deserialize, Clone, Serialize, Debug, Default)]
pub struct Review {
    // When (unix time) did the review take place?
    #[serde(with = "serde_duration_as_secs")]
    pub timestamp: Duration,
    // Recall grade.
    pub grade: Grade,
    // How long you spent before attempting recall.
    #[serde(with = "serde_duration_as_secs")]
    pub time_spent: Duration,
}

impl Review {
    fn new(grade: Grade) -> Self {
        Self {
            timestamp: current_time(),
            grade,
            ..Default::default()
        }
    }
}

#[derive(Deserialize, Clone, Serialize, Debug, Default)]
pub struct Side {
    pub text: String,
    #[serde(flatten)]
    pub audio: AudioSource,
    //#[serde(deserialize_with = "deserialize_image_path")]
    //pub image: ImagePath,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
pub struct Meta {
    pub id: Id,
    pub dependencies: Vec<Id>,
    pub dependents: Vec<Id>,
    pub suspended: bool,
    pub finished: bool,
    pub stability: Option<f32>,
    pub tags: Vec<String>,
}

impl Default for Meta {
    fn default() -> Self {
        Self {
            id: Uuid::new_v4(),
            dependencies: vec![],
            dependents: vec![],
            suspended: false,
            finished: true,
            stability: None,
            tags: vec![],
        }
    }
}

#[cfg(test)]
mod tests {
    use uuid::uuid;

    use super::*;

    #[test]
    fn test_load_cards_from_folder() {
        let _category = Category(vec!["maths".into(), "calculus".into()]);

        //let cards = Card::load_cards_from_folder(&category);
        //insta::assert_debug_snapshot!(cards);
    }

    #[test]
    fn test_card_roundtrip() {
        let mut card = Card::default();
        card.meta.id = uuid!("000a0a00-c943-4c4b-b7bf-f7d483208eb0");
        let category = Category(vec![]);
        let path = card.save_card_to_toml(&category).unwrap();
        let card = Card::parse_toml_to_card(path.as_path());
        insta::assert_debug_snapshot!(card);
    }

    #[test]
    fn test_strength() {
        let stability = 1.0;
        let days_passed = 0.0;
        let recall_rate = Card::calculate_strength(days_passed, stability);
        assert_eq!(recall_rate, 1.0);

        let days_passed = 1.;
        let recall_rate = Card::calculate_strength(days_passed, stability);
        assert_eq!(recall_rate, 0.9);
    }
}
