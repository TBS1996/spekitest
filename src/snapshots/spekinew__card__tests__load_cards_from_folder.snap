---
source: src/card.rs
expression: cards
---
[
    Card {
        front: Side {
            text: "calc q",
            audio: None,
            image: None,
        },
        back: Side {
            text: "calc a",
            audio: None,
            image: None,
        },
        meta: Meta {
            id: 491f8b92-c943-4c4b-b7bf-f7d483208eb0,
            dependencies: [],
            dependents: [],
            suspended: false,
            finished: true,
            tags: [],
        },
        history: [],
    },
    Card {
        front: Side {
            text: "anudda calc",
            audio: None,
            image: None,
        },
        back: Side {
            text: "hey",
            audio: None,
            image: None,
        },
        meta: Meta {
            id: af618145-3da6-48f9-b45c-8d775fc8c7ee,
            dependencies: [],
            dependents: [],
            suspended: false,
            finished: true,
            tags: [],
        },
        history: [],
    },
]
